src/txmempool.h

```diff
@@ -534,6 +534,7 @@ class CTxMemPool
     typedef std::map<uint256, std::vector<CSpentIndexKey> > mapSpentIndexInserted;
     mapSpentIndexInserted mapSpentInserted;
 
+    std::multimap<uint256, uint256> mapProTxRefs; // proTxHash -> transaction (all TXs that refer to an existing proTx)
```

Comment:
Hmm... I wonder if we should protect it (and all these other new maps) with mempool's `cs` lock in `Add/RemoveSmth`.

Reply:
All places where this map is accessed are already protected by `cs`. Locking is either in the method itself (e.g. `addUnchecked`, `existsProviderTxConflict`) or by one of the callers (e.g. `RemoveStaged`, `removeForBlock`)

Reply:
This also applies to all the other maps